vortex-ipc/src/codecs/message_reader/futures.rs

```diff
@@ -0,0 +1,190 @@
+#![cfg(feature = "futures")]
+use std::io;
+
+use bytes::BytesMut;
+use flatbuffers::{root, root_unchecked};
+use futures_util::{AsyncRead, AsyncReadExt};
+use vortex_buffer::Buffer;
+use vortex_error::{vortex_bail, vortex_err, VortexResult};
+
+use crate::codecs::message_reader::MessageReader;
+use crate::flatbuffers::ipc::Message;
+
+pub struct AsyncReadMessageReader<R: AsyncRead + Unpin> {
+    // TODO(ngates): swap this for our own mutable aligned buffer so we can support direct reads.
+    read: R,
+    message: BytesMut,
+    prev_message: BytesMut,
+    finished: bool,
+}
+
+impl<R: AsyncRead + Unpin> AsyncReadMessageReader<R> {
+    pub async fn try_new(read: R) -> VortexResult<Self> {
+        let mut reader = Self {
+            read,
+            message: BytesMut::new(),
+            prev_message: BytesMut::new(),
+            finished: false,
+        };
+        reader.load_next_message().await?;
+        Ok(reader)
+    }
+
+    async fn load_next_message(&mut self) -> VortexResult<bool> {
+        let mut len_buf = [0u8; 4];
+        match self.read.read_exact(&mut len_buf).await {
+            Ok(()) => {}
+            Err(e) => {
+                return match e.kind() {
+                    io::ErrorKind::UnexpectedEof => Ok(false),
+                    _ => Err(e.into()),
+                };
+            }
+        }
+
+        let len = u32::from_le_bytes(len_buf);
+        if len == u32::MAX {
+            // Marker for no more messages.
+            return Ok(false);
+        } else if len == 0 {
+            vortex_bail!(InvalidSerde: "Invalid IPC stream")
+        }
+
+        let mut message = BytesMut::zeroed(len as usize);
```

Comment:
I always thought that since you're going to read into this memory you don't need to zero it. I know that it's very code specific whether it matters ot not

Reply:
There's no way to do that without unsafe, but we reserve + set_len

Reply:
Will FLUP this one